synapse/storage/databases/main/devices.py

```diff
@@ -302,11 +304,17 @@ async def get_device_updates_by_remote(
 
             last_processed_stream_id = update_stream_id
 
+        # len(query_map) + len(cross_signing_keys_by_user) <= len(updates) here,
```

Comment:
Agreed. In the worst case scenario, each update is for a distinct user and is added to either query+map or cross_signing_keys_by_user not both.